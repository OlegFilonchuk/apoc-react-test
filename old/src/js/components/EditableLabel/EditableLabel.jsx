import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import Input from './components/Input';

import './EditableLabel.less';

const EDITABLE_LABEL_CLASS = 'editable-label';

function EditableLabel(props) {
  const { className, inEditMode, children, hasError } = props;

  const wrapperClassName = classNames(EDITABLE_LABEL_CLASS, className, {
    [`${EDITABLE_LABEL_CLASS}--edit`]: inEditMode,
    [`${EDITABLE_LABEL_CLASS}--error`]: hasError
  });

  const inputElement = inEditMode ? <Input {...props} /> : null;

  return (
    <div className={wrapperClassName}>
      <label onClick={props.onClick}>{children}</label>
      {inputElement}
    </div>
  );
}

EditableLabel.propTypes = {
  /**
   * Custom class name for editable element
   */
  className: PropTypes.string,

  /**
   * Label text
   */
  children: PropTypes.string.isRequired,

  /**
   * Is the component in edit mode
   */
  inEditMode: PropTypes.bool,

  /**
   * If element is in error mode
   */
  hasError: PropTypes.bool,

  /**
   * Fired when user clicks on label
   */
  onClick: PropTypes.func
};

EditableLabel.defaultProps = {
  className: '',
  readOnly: false,
  inEditMode: false,
  hasError: false,
  maxLength: 140,
  onClick: () => true
};

export default EditableLabel;
