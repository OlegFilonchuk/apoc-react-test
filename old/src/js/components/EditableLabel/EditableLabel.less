@import '~chameleon/less/variables.less';

.editable-label {
  label {
    display: block;
  }
  input {
    display: none;
  }

  &&--edit {
    label {
      display: none;
    }
    input {
      display: block;
    }
  }

  div&&--error {
    border-color: @error;
  }
}
