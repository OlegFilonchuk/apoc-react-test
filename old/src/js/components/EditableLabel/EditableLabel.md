# Basic example of handling clicking on label

    const onClick = () => {
      setState({ inEditMode: true });
    }

    const onSubmit = () => {};
    const onCanceled = () => {};

    (
      <EditableLabel onClick={onClick} inEditMode={state.inEditMode} onSubmit={onSubmit} onCanceled={onCanceled}>
        Hello there
      </EditableLabel>
    )

# Example of using onSubmit and onCanceled callbacks

    const initialState = {
      value: 'Initial Value'
    };

    const onClick = () => {
      setState({ inEditMode: true });
    };

    const onSubmit = newValue => {
      console.log(`I was updated to ${newValue}`);

      setState({ value: newValue, inEditMode: false });
    };

    const onCanceled = newValue => {
      console.log(`I was about to be changed to ${newValue} but user canceled`);

      setState({ inEditMode: false });
    };

    (
      <EditableLabel onClick={onClick} inEditMode={state.inEditMode} onSubmit={onSubmit} onCanceled={onCanceled}>
        {state.value}
      </EditableLabel>
    )


# Examples of changing props from outside
    const onSubmit = () => true;
    const onCanceled = () => true;

    const initialState  = {
      inEditMode: false,
      counter: 0,
      value: 'Counter: 0'
    };

    const toggleEditMode = () => {
      const { inEditMode } = state;

      setState({ inEditMode: !inEditMode })
    };

    const increment = () => {
      const { counter } = state;
      const newCounter = counter + 1;
      const newValue = `Counter: ${newCounter}`;

      setState({ counter: newCounter, value: newValue });
    };

    const disable = () => {
      setState({ readOnly: true });
    };

    const enable = () => {
      setState({ readOnly: false });
    };

    const addError = () => {
      setState({ hasError: true });
    };

    (
      <div>
        <button onClick={toggleEditMode}>Toggle Edit</button>
        <button onClick={increment}>Increment</button>
        <button onClick={disable}>Disable</button>
        <button onClick={enable}>Enable</button>
        <button onClick={addError}>Add error class</button>
        <EditableLabel onSubmit={onSubmit} onCanceled={onCanceled} inEditMode={state.inEditMode} readOnly={state.readOnly} hasError={state.hasError}>
          { state.value }
        </EditableLabel>
      </div>
    )

# Example of sync validation
It will not submit empty value and '42'

    const initialState = {
      lastValidValue: 'Initial Value',
      value: 'Initial Value',
      hasError: false,
      errorMsg: ''
    };

    const onClick = () => {
      setState({ inEditMode: true });
    };

    const onSubmit = (value) => {
      if (value === state.lastValidValue) {
        console.log('Nothing has changed no validation needed');
        setState({ inEditMode: false });

        return;
      }

      if (value === '') {
        setState({ value, hasError: true, errorMsg: 'Should not be empty' });

        return;
      }

      if (value === '42') {
        setState({ value, hasError: true, errorMsg: 'Should not be equal to 42' });

        return;
      }

      setState({ value, lastValidValue: value,  hasError: false, inEditMode: false });
    };

    const onCanceled = () => {
      setState({ value: state.lastValidValue, inEditMode: false, hasError: false });
    };

    const renderError = () => state.hasError ? <span>{state.errorMsg}</span> : null;

    (
      <div>
        <EditableLabel onClick={onClick} inEditMode={state.inEditMode} onSubmit={onSubmit} onCanceled={onCanceled}>
         {state.value}
        </EditableLabel>
        { renderError() }
      </div>
    )

# Example of async validation
It will not submit empty value and '42'

    const initialState = {
      lastValidValue: 'Initial Value',
      value: 'Initial Value',
      hasError: false,
      errorMsg: ''
    };

    const onClick = () => {
      setState({ inEditMode: true });
    };

    const validate = (value) => {
      if (value === state.lastValidValue) {
        setState({ inEditMode: false, hasError: false, readOnly: false });

        return;
      }

      setTimeout(() => {
        let errorMsg;

        if (value === '') {
         errorMsg = 'Should not be empty';
        } else if (value === '42') {
         errorMsg = 'Should not be equal to 42';
        } else {
          setState({ value , lastValidValue: value, hasError: false, errorMsg, readOnly: false, inEditMode: false });

          return;
        }

        setState({ value , hasError: true, errorMsg, readOnly: false, inEditMode: true });

      }, 2000);
    }

    const onSubmit = value => {
      setState({ readOnly: true });
      validate(value);
    };

    const onCanceled = () => {
      setState({ value: state.lastValidValue, inEditMode: false, hasError: false });
    };

    const renderError = () => state.hasError ? <span>{state.errorMsg}</span> : null;

    (
      <div>
        <EditableLabel onClick={onClick} inEditMode={state.inEditMode} onSubmit={onSubmit} onCanceled={onCanceled} readOnly={state.readOnly}>
         {state.value}
        </EditableLabel>
        { renderError() }
      </div>
    )
