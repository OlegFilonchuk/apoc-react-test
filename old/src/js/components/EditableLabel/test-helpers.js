const KEY_CODES = {
  Enter: 13,
  Escape: 27
};

export const findInput = el => el.find('input');

export const setInputValue = (input, value) => {
  /* eslint-disable no-param-reassign */
  input.simulate('input', { target: { value } });
  /* eslint-enable no-param-reassign */

  return input;
};

export const simulateEvent = (input, eventType, key) => {
  const options = key
    ? {
        key,
        keyCode: KEY_CODES[key],
        which: KEY_CODES[key]
      }
    : {};

  input.simulate(eventType, options);
};
