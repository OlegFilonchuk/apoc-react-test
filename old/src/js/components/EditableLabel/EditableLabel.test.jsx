import React from 'react';
import ReactTestUtils from 'react-dom/test-utils';
import { mount, shallow } from 'enzyme';

import EditableLabel from './EditableLabel';
import { setInputValue, simulateEvent, findInput } from './test-helpers';

describe('<EditableLabel />', () => {
  const initialProps = {
    onSubmit: jest.fn(),
    onCanceled: jest.fn()
  };
  const initalValue = 'some initial value';
  const mountEditableLabel = (props = {}, mountFn = mount) =>
    mountFn(
      <EditableLabel {...initialProps} {...props}>
        {initalValue}
      </EditableLabel>
    );
  const newValue = 'some new value';

  describe('should render correct structure', () => {
    it('default', () => {
      const editableLabel = mountEditableLabel({}, shallow);

      expect(editableLabel).toMatchSnapshot();
    });

    it('with additional class name passed as props', () => {
      const className = 'custom-class';
      const editableLabel = mountEditableLabel({ className }, shallow);

      expect(editableLabel).toHaveClassName(className);
    });

    it('in edit mode', () => {
      const editableLabel = mountEditableLabel({ inEditMode: true }, shallow);

      expect(editableLabel).toMatchSnapshot();
    });

    it('read only', () => {
      const editableLabel = mountEditableLabel({ readOnly: true }, shallow);

      expect(editableLabel).toMatchSnapshot();
    });

    it('with error', () => {
      const editableLabel = mountEditableLabel({ hasError: true }, shallow);

      expect(editableLabel).toMatchSnapshot();
    });
  });

  describe('onClick', () => {
    it('should fire onClick event when clicking on label', () => {
      const onClick = jest.fn();
      const editableLabel = mountEditableLabel({ onClick });
      const label = editableLabel.find('label');

      simulateEvent(label, 'click');

      expect(onClick).toHaveBeenCalled();
    });
  });

  describe('onSubmit simple check', () => {
    it('should fire onSubmit callback after entering value and confirming with enter', () => {
      const onSubmit = jest.fn();
      const editableLabel = mountEditableLabel({ inEditMode: true, onSubmit });
      const input = findInput(editableLabel).getDOMNode();

      input.value = newValue;

      ReactTestUtils.Simulate.keyUp(input, { key: 'Enter', keyCode: 13, which: 13 });

      expect(onSubmit).toHaveBeenCalledWith(newValue);
    });

    it('should fire onSubmit callback after entering value and confirming with blur', () => {
      const onSubmit = jest.fn();
      const editableLabel = mountEditableLabel({ inEditMode: true, onSubmit });
      const input = findInput(editableLabel).getDOMNode();

      input.value = newValue;

      ReactTestUtils.Simulate.blur(input);

      expect(onSubmit).toHaveBeenCalledWith(newValue);
    });

    it('should not fire onSubmit callback after entering value and canceling with escape', () => {
      const onSubmit = jest.fn();
      const editableLabel = mountEditableLabel({ inEditMode: true, onSubmit });
      const input = findInput(editableLabel);

      setInputValue(input, newValue);

      simulateEvent(input, 'keyup', 'Escape');

      expect(onSubmit).not.toHaveBeenCalled();
    });
  });

  describe('onCanceled', () => {
    it('should fire onCanceled callback after entering value and canceling', () => {
      // given
      const onCanceled = jest.fn();
      const editableLabel = mountEditableLabel({ inEditMode: true, onCanceled });
      const input = findInput(editableLabel);

      // when
      setInputValue(input, newValue);
      simulateEvent(input, 'keyup', 'Escape');

      // then
      expect(onCanceled).toHaveBeenCalled();
    });

    it('should not fire onCanceled callback after entering value and confirming', () => {
      // given
      const onCanceled = jest.fn();
      const editableLabel = mountEditableLabel({ inEditMode: true, onCanceled });
      const input = findInput(editableLabel);

      // when
      setInputValue(input, newValue);
      simulateEvent(input, 'keyup', 'Enter');

      // then
      expect(onCanceled).not.toHaveBeenCalled();
    });
  });
});
