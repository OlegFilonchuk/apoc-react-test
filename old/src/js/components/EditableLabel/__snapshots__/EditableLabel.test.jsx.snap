// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditableLabel /> should render correct structure default 1`] = `
<div
  className="editable-label"
>
  <label
    onClick={[Function]}
  >
    some initial value
  </label>
</div>
`;

exports[`<EditableLabel /> should render correct structure in edit mode 1`] = `
<div
  className="editable-label editable-label--edit"
>
  <label
    onClick={[Function]}
  >
    some initial value
  </label>
  <Input
    alreadyFocused={false}
    className=""
    hasError={false}
    inEditMode={true}
    maxLength={140}
    onCanceled={[Function]}
    onChange={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onSubmit={[Function]}
    readOnly={false}
  >
    some initial value
  </Input>
</div>
`;

exports[`<EditableLabel /> should render correct structure read only 1`] = `
<div
  className="editable-label"
>
  <label
    onClick={[Function]}
  >
    some initial value
  </label>
</div>
`;

exports[`<EditableLabel /> should render correct structure with error 1`] = `
<div
  className="editable-label editable-label--error"
>
  <label
    onClick={[Function]}
  >
    some initial value
  </label>
</div>
`;
